File: frontend/src/schedules.ts
import { serverNow } from './serverClock.ts'

// Types and pure presentation helpers for the Schedule tab.
//
// Everything here is a function of data the daemon already sent. The one thing the
// browser deliberately does NOT do is compute a next fire time: cron plus a timezone
// plus daylight saving is exactly the arithmetic that must have one implementation,
// and it lives in `schedules.py`. The editor asks `/api/schedules/preview` instead,
// so what the user is shown before saving is produced by the same code that will
// actually fire it.

export type ScheduleTrigger = 'cron' | 'interval' | 'once'
export type ScheduleOverlap = 'skip' | 'allow'

/** What the trigger is attached to. */
export type ScheduleAction = 'spawn' | 'resume'

/** Which conversation a resume reopens, and how the answer is allowed to move.
 *
 *  Three, because "resume this session later" has three genuinely different answers
 *  and picking one for the author would be picking wrong. See `TARGET_KIND_COPY`. */
export type ScheduleTargetKind = 'run' | 'latest_of_session' | 'fork_point'

/** What the daemon says a resume schedule currently points at.
 *
 *  Resolved on read, never stored beside the id: a label kept with the schedule would
 *  go on naming a conversation that has since been deleted, and `missing` is the
 *  honest answer to that. `resolved_*` differs from the pinned row only for a rolling
 *  continuation, which is the one kind whose target moves. */
export type ScheduleTarget = {
  run_id: string
  missing: boolean
  backend?: string
  name?: string
  resolved_run_id?: string
  resolved_name?: string
  resolved_at?: number | null
  context_pct?: number | null
}

export type ScheduleFollowUp = {
  body: string
  delay_seconds: number
  armed: boolean
}

export type ScheduleRun = {
  id: string
  schedule_id: string
  project_id: string
  fire_key: string
  due_at: number | null
  started_at: number
  finished_at: number | null
  outcome: 'started' | 'spawned' | 'skipped' | 'failed' | 'missed'
  reason: string
  session_id: string | null
  origin: 'timer' | 'manual'
  queued_messages: number
}

export type Schedule = {
  id: string
  project_id: string
  project_name: string
  label: string
  enabled: boolean
  trigger_kind: ScheduleTrigger
  cron: string
  interval_seconds: number | null
  run_at: number | null
  timezone: string
  catch_up: boolean
  overlap: ScheduleOverlap
  backend: string
  profile_id: string
  cwd: string
  session_name: string
  prompt: string
  follow_ups: ScheduleFollowUp[]
  daily_run_cap: number
  next_fire_at: number | null
  last_fire_at: number | null
  last_session_id: string | null
  last_outcome: string
  last_reason: string
  disabled_reason: string
  /** Live permission answer, not a stored flag. Empty means it can fire. */
  blocked: '' | 'project_missing' | 'automation_disabled' | 'install_disabled'
  runs: ScheduleRun[]
  revision: number
  created_at: number
  updated_at: number
  action: ScheduleAction
  target_run_id: string
  target_kind: ScheduleTargetKind
  target_cut_message_id: string
  target_cut_mode: '' | 'before' | 'after'
  context_ceiling_pct: number
  /** Null on a spawn, which has nothing to point at. */
  target: ScheduleTarget | null
}

export type ScheduleStatus = {
  enabled: boolean
  total: number
  armed: number
  live_sessions: number
  max_concurrent: number
}

export type ScheduleDraft = {
  label: string
  prompt: string
  trigger_kind: ScheduleTrigger
  cron: string
  /** Minutes in the editor; seconds on the wire. A field measured in seconds
   *  invites `3600` typed as `360`. */
  interval_minutes: number
  run_at_local: string
  timezone: string
  catch_up: boolean
  overlap: ScheduleOverlap
  backend: string
  profile_id: string
  session_name: string
  daily_run_cap: number
  follow_ups: string[]
  action: ScheduleAction
  target_run_id: string
  target_kind: ScheduleTargetKind
  target_cut_message_id: string
  target_cut_mode: '' | 'before' | 'after'
  /** 0-1. Only a rolling continuation reads it. */
  context_ceiling_pct: number
  /** Display only, never sent: what the target conversation is called, so the editor
   *  can name what it is about to reopen without a second fetch. */
  target_label: string
  target_backend: string
}

/** A named cron expression, and the piece of the syntax it demonstrates.
 *
 *  The list is short and each entry earns its place twice: it is a rhythm people
 *  actually want, and between them they cover every part of the five-field grammar -
 *  fixed values, names, ranges, lists, steps, and the day-of-month field. Picking one
 *  fills the field, so the next thing a person does is edit a working expression
 *  rather than write one from a blank box.
 *
 *  `teaches` is shown under the row for whichever preset the current expression
 *  matches, which is why matching is by exact string: a paraphrase of what the user
 *  typed would be a second, weaker cron implementation in the browser. */
export type CronPreset = { label: string; cron: string; teaches: string }

export const CRON_PRESETS: CronPreset[] = [
  {
    label: 'Every day at 09:00',
    cron: '0 9 * * *',
    teaches: 'The five fields are minute, hour, day-of-month, month, day-of-week. Fix the first two, leave the rest as * for "every".',
  },
  {
    label: 'Every weekday at 09:00',
    cron: '0 9 * * mon-fri',
    teaches: 'The last field is the weekday, and it takes names and ranges: mon-fri is Monday through Friday.',
  },
  {
    label: 'Every Monday at 09:00',
    cron: '0 9 * * mon',
    teaches: 'Weekly is just one weekday. Pin the day-of-week field and leave day-of-month as *.',
  },
  {
    label: 'Every Wednesday at 13:00',
    cron: '0 13 * * wed',
    teaches: 'Cron counts days and months, never weeks, so there is no "every other Wednesday". For a fortnightly rhythm use the 1st and 15th below, or an interval trigger.',
  },
  {
    label: 'The 1st and 15th at 13:00',
    cron: '0 13 1,15 * *',
    teaches: 'A comma is a list: 1,15 is the first and fifteenth of the month. This is the closest cron gets to every other week.',
  },
  {
    label: 'Every 30 minutes',
    cron: '*/30 * * * *',
    teaches: 'A slash is a step: */30 in the minute field means every 30th minute, and the same works in any field.',
  },
  {
    label: 'First of the month at 03:00',
    cron: '0 3 1 * *',
    teaches: 'The third field is the day of the month. With a weekday of * this fires on that date whatever day it lands on.',
  },
]

/** The preset one expression matches exactly, if any. Whitespace-tolerant, because
 *  the daemon normalizes the expression it stores and the field is free text. */
export function presetForCron(cron: string): CronPreset | undefined {
  const normalized = cron.trim().split(/\s+/).join(' ').toLowerCase()
  return CRON_PRESETS.find(preset => preset.cron === normalized)
}

/** The three target kinds, said in the words of what each one does and costs.
 *
 *  The trade-off is the whole content of the choice, so it is stated on the option
 *  rather than hidden in a doc: a pinned run cannot drift, a rolling continuation
 *  accumulates until compaction eats its own early context, and a pinned point is the
 *  only one a recurring trigger repeats without run N+1 inheriting run N. */
export const TARGET_KIND_COPY: Record<ScheduleTargetKind, { label: string; detail: string }> = {
  run: {
    label: 'This exact conversation',
    detail: 'Reopens the conversation as it is now. Nothing drifts. Best for a one-off.',
  },
  latest_of_session: {
    label: 'Wherever this work has got to',
    detail: 'Follows the conversation through rollovers and earlier resumes, so each run continues the last. It keeps growing: every resume replays the whole conversation, and past the ceiling below its early context has already been compacted into a summary.',
  },
  fork_point: {
    label: 'From a fixed point, every time',
    detail: 'Forks the conversation at a message you pick and resumes the copy, leaving the original untouched. Every run starts from identical context, which is what makes a repeating trigger repeatable. Claude only.',
  },
}

export const BLOCKED_LABELS: Record<string, string> = {
  project_missing: 'this Project is no longer registered',
  automation_disabled: 'Scheduled runs is off for this Project',
  install_disabled: 'scheduled runs are switched off for this install',
}

export const OUTCOME_LABELS: Record<string, string> = {
  started: 'starting',
  spawned: 'ran',
  skipped: 'skipped',
  failed: 'failed',
  missed: 'missed',
}

export const emptyDraft = (): ScheduleDraft => ({
  label: '',
  prompt: '',
  trigger_kind: 'cron',
  cron: '0 3 * * *',
  interval_minutes: 360,
  run_at_local: '',
  timezone: '',
  catch_up: false,
  overlap: 'skip',
  backend: '',
  profile_id: '',
  session_name: '',
  daily_run_cap: 0,
  follow_ups: [],
  action: 'spawn',
  target_run_id: '',
  target_kind: 'run',
  target_cut_message_id: '',
  target_cut_mode: '',
  context_ceiling_pct: 0,
  target_label: '',
  target_backend: '',
})

/** The conversation a resume schedule is seeded from, as the callers can describe it.
 *
 *  `run_id` is a **history run id**, never a session id, and both call sites have to
 *  pass one: History rows are keyed by it, and a live pane's is `agent_run_id` rather
 *  than its own id, because a pane that rolled its conversation owns a row keyed by
 *  the run. */
export type ResumeSeed = {
  run_id: string
  label: string
  backend: string
  /** Pre-selects the kind. A live pane seeds "wherever it has got to"; a History row
   *  seeds the pinned conversation, which is what the reader is looking at. */
  kind: ScheduleTargetKind
}

/** A one-off resume of `seed`, ready for the editor.
 *
 *  Defaults to `once` with an empty time rather than to a cron expression: the reason
 *  somebody reaches this from a conversation is almost always "pick this up on
 *  Tuesday", and offering `0 3 * * *` would arm a nightly agent by accident. */
export function resumeDraft(seed: ResumeSeed): ScheduleDraft {
  return {
    ...emptyDraft(),
    label: seed.label.slice(0, 80),
    trigger_kind: 'once',
    cron: '0 9 * * *',
    action: 'resume',
    target_run_id: seed.run_id,
    target_kind: seed.kind,
    context_ceiling_pct: seed.kind === 'latest_of_session' ? 0.7 : 0,
    target_label: seed.label,
    target_backend: seed.backend,
  }
}

export function draftFromSchedule(schedule: Schedule): ScheduleDraft {
  return {
    label: schedule.label,
    prompt: schedule.prompt,
    trigger_kind: schedule.trigger_kind,
    cron: schedule.cron || '0 3 * * *',
    interval_minutes: schedule.interval_seconds ? Math.round(schedule.interval_seconds / 60) : 360,
    run_at_local: schedule.run_at ? localInputValue(schedule.run_at) : '',
    timezone: schedule.timezone,
    catch_up: schedule.catch_up,
    overlap: schedule.overlap,
    backend: schedule.backend,
    profile_id: schedule.profile_id,
    session_name: schedule.session_name,
    daily_run_cap: schedule.daily_run_cap,
    follow_ups: schedule.follow_ups.map(item => item.body),
    action: schedule.action,
    target_run_id: schedule.target_run_id,
    target_kind: schedule.target_kind,
    target_cut_message_id: schedule.target_cut_message_id,
    target_cut_mode: schedule.target_cut_mode,
    context_ceiling_pct: schedule.context_ceiling_pct,
    target_label: schedule.target?.name || '',
    target_backend: schedule.target?.backend || '',
  }
}

/** `datetime-local` wants the browser's own wall clock, without a zone suffix. */
export function localInputValue(epochSeconds: number): string {
  const moment = new Date(epochSeconds * 1000)
  const pad = (value: number) => String(value).padStart(2, '0')
  return `${moment.getFullYear()}-${pad(moment.getMonth() + 1)}-${pad(moment.getDate())}`
    + `T${pad(moment.getHours())}:${pad(moment.getMinutes())}`
}

export function draftToBody(draft: ScheduleDraft): Record<string, unknown> {
  const resuming = draft.action === 'resume'
  const body: Record<string, unknown> = {
    label: draft.label.trim(),
    prompt: draft.prompt,
    trigger_kind: draft.trigger_kind,
    timezone: draft.timezone.trim(),
    catch_up: draft.catch_up,
    // A conversation opens once, so a resume has no overlap policy to choose and the
    // daemon refuses one. Sent as the only legal value rather than omitted, so a draft
    // that was a spawn a moment ago cannot carry `allow` across the switch.
    overlap: resuming ? 'skip' : draft.overlap,
    session_name: draft.session_name.trim(),
    daily_run_cap: draft.daily_run_cap || 0,
    follow_ups: draft.follow_ups.map(body => body.trim()).filter(Boolean),
    action: draft.action,
  }
  if (resuming) {
    // The harness, its arguments and the working directory all come from the
    // conversation's own History row. Sending any of them is refused rather than
    // ignored, so they are left off entirely.
    body.target_run_id = draft.target_run_id
    body.target_kind = draft.target_kind
    if (draft.target_kind === 'fork_point') {
      body.target_cut_message_id = draft.target_cut_message_id
      body.target_cut_mode = draft.target_cut_mode || 'after'
    }
    body.context_ceiling_pct = draft.target_kind === 'latest_of_session'
      ? draft.context_ceiling_pct
      : 0
  } else {
    body.backend = draft.backend
    body.profile_id = draft.profile_id
  }
  if (draft.trigger_kind === 'cron') body.cron = draft.cron.trim()
  if (draft.trigger_kind === 'interval') body.interval_seconds = Math.round(draft.interval_minutes * 60)
  if (draft.trigger_kind === 'once') {
    const parsed = draft.run_at_local ? Date.parse(draft.run_at_local) : NaN
    body.run_at = Number.isNaN(parsed) ? 0 : parsed / 1000
  }
  return body
}

/** What a row does, in one phrase: the half of a schedule the cadence does not say. */
export function actionLabel(schedule: Schedule): string {
  if (schedule.action !== 'resume') return 'starts a new session'
  const target = schedule.target
  if (!target || target.missing) return 'resumes a conversation that is gone'
  if (schedule.target_kind === 'fork_point') return `forks ${target.name || 'a conversation'}`
  if (schedule.target_kind === 'latest_of_session') {
    // Names where the conversation actually is, not where it was pinned. That is the
    // one thing about a rolling target a reader cannot work out from the row.
    return `continues ${target.resolved_name || target.name || 'a conversation'}`
  }
  return `resumes ${target.name || 'a conversation'}`
}

/** Whether this schedule needs a conversation to exist, and it does not. */
export const targetIsMissing = (schedule: Schedule): boolean =>
  schedule.action === 'resume' && !!schedule.target?.missing

/** What the row says under the label: the trigger, in words. */
export function cadenceLabel(schedule: Schedule | ScheduleDraft): string {
  const kind = schedule.trigger_kind
  if (kind === 'cron') {
    const expression = 'cron' in schedule ? schedule.cron : ''
    const zone = schedule.timezone ? ` (${schedule.timezone})` : ''
    return `cron ${expression}${zone}`
  }
  if (kind === 'interval') {
    const seconds = 'interval_seconds' in schedule
      ? Number(schedule.interval_seconds || 0)
      : Math.round(schedule.interval_minutes * 60)
    return `every ${durationLabel(seconds)}`
  }
  const at = 'run_at' in schedule ? schedule.run_at : null
  return at ? `once, ${absoluteLabel(at)}` : 'once'
}

export function durationLabel(seconds: number): string {
  if (!seconds || seconds < 0) return '—'
  if (seconds < 3600) return `${Math.round(seconds / 60)} min`
  if (seconds < 86_400) {
    const hours = seconds / 3600
    return `${Number.isInteger(hours) ? hours : hours.toFixed(1)} h`
  }
  const days = seconds / 86_400
  return `${Number.isInteger(days) ? days : days.toFixed(1)} d`
}

/** "in 4h" / "in 3d" / "now" for a future instant, using the daemon's clock. */
export function untilLabel(epochSeconds: number | null, now: number = serverNow()): string {
  if (!epochSeconds) return 'not scheduled'
  const delta = epochSeconds - now
  if (delta <= 0) return 'due'
  if (delta < 90) return 'in under a minute'
  if (delta < 3600) return `in ${Math.round(delta / 60)}m`
  if (delta < 86_400) return `in ${Math.round(delta / 3600)}h`
  return `in ${Math.round(delta / 86_400)}d`
}

export function absoluteLabel(epochSeconds: number | null): string {
  if (!epochSeconds) return '—'
  return new Date(epochSeconds * 1000).toLocaleString([], {
    month: 'short', day: 'numeric', hour: '2-digit', minute: '2-digit',
  })
}

/** The one-line verdict a row leads with. */
export function outcomeLabel(schedule: Schedule): string {
  if (!schedule.last_outcome) return 'never run'
  const label = OUTCOME_LABELS[schedule.last_outcome] || schedule.last_outcome
  return `${label} ${absoluteLabel(schedule.last_fire_at)}`
}

/** Rows that need a human: a failure, or armed-but-cannot-fire.
 *
 *  This is what the rail badge counts. A paused schedule is deliberately not
 *  included - pausing is a decision someone already made, and badging it would
 *  train people to ignore the badge. */
export function needsAttention(schedules: Schedule[]): Schedule[] {
  return schedules.filter(schedule =>
    schedule.last_outcome === 'failed'
    || (schedule.enabled && schedule.blocked !== '')
    // An armed resume whose conversation has been deleted can never fire. It is the
    // same class of failure as `blocked` - a row that looks armed and is inert - and
    // it is worth badging *before* the next fire disables it, because the fire may be
    // a week away.
    || (schedule.enabled && targetIsMissing(schedule)))
}

/** Sort for display: armed and soonest first, then paused, then unarmed. */
export function orderSchedules(schedules: Schedule[]): Schedule[] {
  return [...schedules].sort((left, right) => {
    if (left.enabled !== right.enabled) return left.enabled ? -1 : 1
    const leftFire = left.next_fire_at ?? Number.POSITIVE_INFINITY
    const rightFire = right.next_fire_at ?? Number.POSITIVE_INFINITY
    if (leftFire !== rightFire) return leftFire - rightFire
    return left.label.localeCompare(right.label)
  })
}

File: src/swe_mux/routes/schedules.py
"""Scheduled runs: definitions, previews, and their run history."""

from __future__ import annotations

import logging
import time
from typing import Any

from aiohttp import web

from .. import (
    app_keys as keys,
)
from .. import (
    session_titles,
)
from ..config import Config
from ..http_support import json_response
from ..schedule_store import ScheduleStore
from ..scheduler import ScheduleService, spec_from_row
from ..schedules import first_occurrence, next_occurrence, parse_spec
from ..session_resume import resolve_latest_run
from .support import _observations_project

log = logging.getLogger(__name__)


def _schedule_service(request: web.Request) -> ScheduleService:
    service = request.app.get(keys.SCHEDULES)
    if service is None:  # pragma: no cover - only a partially built app
        raise web.HTTPServiceUnavailable(text="scheduled runs are unavailable")
    return service


async def _schedule_view(
    request: web.Request, schedule: dict[str, Any], *, runs: int = 5
) -> dict[str, Any]:
    """One schedule, plus the two things a reader cannot derive from the row.

    `blocked` is the live permission answer rather than a stored flag: a Project
    can be opted out after a schedule was written, and a row that still reads
    `enabled` while nothing will ever fire is the exact lie this surface exists
    to avoid. `runs` is the recent history the tab shows under the row.
    """
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    project = request.app[keys.PROJECTS].projects.get(str(schedule["project_id"]))
    blocked = ""
    if project is None:
        blocked = "project_missing"
    else:
        gate = request.app.get(keys.AUTOMATION_GATE)
        if gate is not None and "scheduled_runs" not in await gate(str(project.root)):
            blocked = "automation_disabled"
    config: Config = request.app[keys.CONFIG]
    if not config.scheduled_runs_enabled:
        blocked = blocked or "install_disabled"
    return {
        **schedule,
        "project_name": project.name if project else "",
        "blocked": blocked,
        "target": await _schedule_target_view(request, schedule),
        "runs": await store.runs(str(schedule["id"]), limit=runs),
    }


async def _schedule_target_view(
    request: web.Request, schedule: dict[str, Any]
) -> dict[str, Any] | None:
    """What a resume schedule points at, named the way the operator would name it.

    Resolved on read rather than stored beside the id, for the same reason `blocked`
    is: a stored label would keep claiming a conversation that has since been deleted
    from History, and a row that reads armed against a target that no longer exists is
    exactly the lie this surface exists to avoid. ``missing`` is the honest answer, and
    the schedule's next fire will disable itself on it.

    For a rolling continuation this also reports where the conversation has actually
    got to, which is the one thing about that kind a reader cannot work out for
    themselves.
    """
    if str(schedule.get("action") or "spawn") != "resume":
        return None
    run_id = str(schedule.get("target_run_id") or "")
    history = request.app.get(keys.HISTORY)
    if not run_id or history is None:
        return {"run_id": run_id, "missing": True}
    row = await history.history_entry(run_id)
    if row is None:
        return {"run_id": run_id, "missing": True}
    resolved = row
    if str(schedule.get("target_kind")) == "latest_of_session":
        resolved = (
            await resolve_latest_run(
                run_id,
                history=history,
                automation_store=request.app[keys.AUTOMATION_STORE],
            )
            or row
        )
    # The two-name rule is `session_titles.py`'s, and asking it is what keeps this row in
    # step with the History browser, the sidebar, and every other surface that names a run.
    titles = await session_titles.generated_titles(
        request.app.get(keys.AUTOMATION_STORE),
        {session_titles.row_run_id(row), session_titles.row_run_id(resolved)},
    )
    return {
        "run_id": run_id,
        "missing": False,
        "backend": str(row.get("backend") or ""),
        "name": session_titles.row_display_name(row, titles),
        "resolved_run_id": str(resolved["id"]),
        "resolved_name": session_titles.row_display_name(resolved, titles),
        "resolved_at": resolved.get("last_message_at") or resolved.get("spawned_at"),
        "context_pct": resolved.get("final_context_pct"),
    }


async def list_schedules(request: web.Request) -> web.Response:
    """Every schedule, or one Project's.

    The unscoped form is what makes the tab's fleet toggle possible: "what fires
    tonight" is not a per-Project question, even though every schedule belongs to
    exactly one Project.
    """
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    project_id = request.query.get("project_id") or None
    if project_id and project_id not in request.app[keys.PROJECTS].projects:
        raise ValueError(f"unknown project: {project_id}")
    rows = await store.list_schedules(project_id)
    service = _schedule_service(request)
    return json_response(
        {
            "schedules": [await _schedule_view(request, row) for row in rows],
            "status": await service.status(),
        }
    )


async def list_project_schedules(request: web.Request) -> web.Response:
    project = _observations_project(request)
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    rows = await store.list_schedules(project.id)
    service = _schedule_service(request)
    return json_response(
        {
            "project_id": project.id,
            "schedules": [await _schedule_view(request, row) for row in rows],
            "status": await service.status(),
        }
    )


async def create_project_schedule(request: web.Request) -> web.Response:
    """Write a new schedule for one Project.

    Permission is deliberately *not* required to write one: a user may author a
    schedule and opt the Project in afterwards, and refusing the write would make
    the toggle discoverable only by failing. What permission gates is firing, and
    the response says so through `blocked`.
    """
    project = _observations_project(request)
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    body = await request.json()
    spec = parse_spec(body)
    now = time.time()
    schedule = await store.create(
        project_id=project.id,
        project_root=str(project.root),
        spec=spec,
        next_fire_at=first_occurrence(spec, now=now),
        now=now,
    )
    await request.app[keys.EVENTS].emit(
        "schedule_changed",
        source="user",
        action="created",
        schedule_id=schedule["id"],
        project_id=project.id,
    )
    return json_response(await _schedule_view(request, schedule), 201)


async def patch_schedule(request: web.Request) -> web.Response:
    """Replace a definition, or arm/disarm it.

    A body carrying only `enabled` is the pause switch and keeps the existing
    definition; anything else is a full replacement validated exactly like a
    create, because a schedule is small and a partial-update surface over a
    trigger is how one ends up with a cron expression and an interval both set.
    """
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    schedule_id = request.match_info["schedule_id"]
    current = await store.get(schedule_id)
    if current is None:
        raise KeyError(schedule_id)
    body = await request.json()
    revision = body.get("revision")
    if revision is not None and not isinstance(revision, int):
        raise ValueError("revision must be an integer")
    now = time.time()
    if set(body) <= {"enabled", "revision"} and "enabled" in body:
        spec = spec_from_row(current)
        enabled = bool(body["enabled"])
        updated = await store.set_enabled(
            schedule_id,
            enabled,
            next_fire_at=first_occurrence(spec, now=now) if enabled else None,
            reason="" if enabled else "paused",
            now=now,
        )
    else:
        spec = parse_spec(body)
        updated = await store.replace(
            schedule_id,
            spec=spec,
            next_fire_at=first_occurrence(spec, now=now),
            revision=revision,
            now=now,
        )
        if updated is None:
            return json_response(
                {
                    "error": "this schedule changed elsewhere; re-read it and try again",
                    "code": "revision_conflict",
                },
                409,
            )
    if updated is None:
        raise KeyError(schedule_id)
    await request.app[keys.EVENTS].emit(
        "schedule_changed",
        source="user",
        action="updated",
        schedule_id=schedule_id,
        project_id=str(updated["project_id"]),
    )
    return json_response(await _schedule_view(request, updated))


async def delete_schedule(request: web.Request) -> web.Response:
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    schedule_id = request.match_info["schedule_id"]
    existing = await store.get(schedule_id)
    if not await store.delete(schedule_id):
        raise KeyError(schedule_id)
    await request.app[keys.EVENTS].emit(
        "schedule_changed",
        source="user",
        action="deleted",
        schedule_id=schedule_id,
        project_id=str(existing["project_id"]) if existing else "",
    )
    return json_response({"deleted": True, "id": schedule_id})


async def run_schedule_now(request: web.Request) -> web.Response:
    """Fire one schedule immediately.

    Still subject to every fire-time guard except lateness - an explicit request
    is never a missed window - so "Run now" cannot be used to walk around the
    Project opt-in, the overlap policy, or the concurrency ceiling.
    """
    service = _schedule_service(request)
    schedule_id = request.match_info["schedule_id"]
    run = await service.run_now(schedule_id)
    if run is None:
        raise KeyError(schedule_id)
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    schedule = await store.get(schedule_id)
    return json_response(
        {
            "run": run,
            "schedule": await _schedule_view(request, schedule) if schedule else None,
        }
    )


async def list_schedule_runs(request: web.Request) -> web.Response:
    store: ScheduleStore = request.app[keys.SCHEDULE_STORE]
    schedule_id = request.match_info["schedule_id"]
    if await store.get(schedule_id) is None:
        raise KeyError(schedule_id)
    limit = int(request.query.get("limit") or 50)
    return json_response({"runs": await store.runs(schedule_id, limit=limit)})


async def preview_schedule(request: web.Request) -> web.Response:
    """Answer "when would this fire" for an unsaved definition.

    The editor calls this on every trigger change. A cron expression is the one
    field in this feature whose meaning cannot be read off its own text, and a
    surface that shows the next three fire times is the difference between
    writing one confidently and finding out at 3 a.m.
    """
    spec = parse_spec(await request.json())
    now = time.time()
    fires: list[float] = []
    cursor = now
    for _ in range(5):
        following = next_occurrence(spec, cursor)
        if following is None:
            break
        fires.append(following)
        cursor = following
    return json_response({"next_fires": fires, "now": now})


ROUTES: tuple[web.RouteDef, ...] = (
    # Scheduled runs. Definitions are Project-owned (a spawn belongs to
    # exactly one Project) while the listing is also reachable unscoped,
    # because "what fires tonight" spans them.
    web.get("/api/schedules", list_schedules),
    web.post("/api/schedules/preview", preview_schedule),
    web.get("/api/projects/{project_id}/schedules", list_project_schedules),
    web.post("/api/projects/{project_id}/schedules", create_project_schedule),
    web.patch("/api/schedules/{schedule_id}", patch_schedule),
    web.delete("/api/schedules/{schedule_id}", delete_schedule),
    web.post("/api/schedules/{schedule_id}/run", run_schedule_now),
    web.get("/api/schedules/{schedule_id}/runs", list_schedule_runs),
)
